### PR TITLE
Df separate mesh reader

### PR DESCRIPTION
**Pull request by dflanderka**

### CODE REVIEW

File: src/mesh/mesh.cc [hunk 1]
```diff
@@ -773,5 +739,48 @@ const BIHTree &Mesh::get_bih_tree() {
...
+	if ( !region_idx.is_valid() ) {
+		region_idx = region_db_.add_region( region_id, region_db_.create_label_from_id(region_id), dim, "$Element" );
+	}
+	region_db_.mark_used_region(region_idx.idx());
+
+	if (region_idx.is_boundary()) {
+		ele = bc_elements.add_item(elm_id);
+	} else {
+		if(dim == 0 )
+			WarningOut().fmt("Bulk elements of zero size(dim=0) are not supported. Element ID: {}.\n", elm_id);
```
Comment (dflanderka) [inline src/mesh/mesh.cc]: We are potential problem here. If warning is produced, 'ele' object is not initialized and next method 'ele-&gt;init' failed on segfault.

File: src/io/msh_basereader.cc [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+#include "io/msh_vtkreader.hh"
+#include "system/sys_profiler.hh"
+
+
+BaseMeshReader::BaseMeshReader(const Input::Record &mesh_rec)
+: tok_(mesh_rec.val<FilePath>("mesh_file")) {
+	input_mesh_rec_ = mesh_rec;
+}
+
+BaseMeshReader::BaseMeshReader(std::istream &in)
```
Comment (jbrezmorf) [inline src/io/msh_basereader.cc]: - have just single constructor which takes stream or rather FilePath as parameter
- static method `reader factory` can be used to construct reader depending on the extension
- static method `mesh_factory` can be used to construct reader and call read_mesh, the mesh input record is passed to it and to `read_mesh` as well

File: src/mesh/mesh.cc (hunk 1 above)
Comment (jbrezmorf) [inline src/mesh/mesh.cc]: add continue.

File: src/fields/field_elementwise.cc [hunk 3]
```diff
@@ -36,7 +36,7 @@ const Input::Type::Record & FieldElementwise<spacedim, Value>::get_input_type()
     return IT::Record("FieldElementwise", FieldAlgorithmBase<spacedim,Value>::template_name()+" Field constant in space.")
         .derive_from(FieldAlgorithmBase<spacedim, Value>::get_input_type())
         .copy_keys(FieldAlgorithmBase<spacedim, Value>::get_field_algo_common_keys())
-        .declare_key("gmsh_file", IT::FileName::input(), IT::Default::obligatory(),
+        .declare_key("mesh_file", IT::FileName::input(), IT::Default::obligatory(),
```
Comment (jbrezmorf) [inline src/fields/field_elementwise.cc]: revert

File: src/io/msh_basereader.cc [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+	try {
+		if ( file_path.extension() == ".msh" ) {
+			reader_ptr = std::make_shared<GmshMeshReader>(mesh_rec);
+		} else if ( file_path.extension() == ".vtu" ) {
+			reader_ptr = std::make_shared<VtkMeshReader>(mesh_rec);
+		} else {
+			THROW(ExcWrongExtension() << EI_FileExtension(file_path.extension()) << EI_MeshFile((string)file_path) );
+		}
+    } INPUT_CATCH(FilePath::ExcFileOpen, FilePath::EI_Address_String, mesh_rec)
+	catch (ExceptionBase const &e) {
```
Comment (jbrezmorf) [inline src/io/msh_basereader.cc]: ?? can remove

File: src/io/observe.cc [hunk 5]
```diff
@@ -316,19 +319,21 @@ void Observe::compute_field_values(Field<spacedim, Value> &field)
...
         unsigned int ele_index = o_point.element_idx_;
         const Value &obs_value =
-                Value( const_cast<typename Value::return_type &>(
-                        field.value(o_point.global_coords_,
-                                ElementAccessor<spacedim>(this->mesh_, ele_index,false)) ));
-        output_data.store_value(i_data,  obs_value);
+                        Value( const_cast<typename Value::return_type &>(
+                                field.value(o_point.global_coords_,
+                                        ElementAccessor<spacedim>(this->mesh_, ele_index,false)) ));
+        output_data.store_value(i_data,  obs_value.mem_ptr());
```
Comment (jbrezmorf) [inline src/io/observe.cc]: add assert that output_data single element size match size of obs_value, i.e. obs_value.n_rows()*obs_value.n_cols().

Or at least add comment. Simliarly in register field values in OutputTime.

**Comment by dflanderka** [comment]
Observe unit test fixed, build is successful now and branch can be merged.

**Pull request merged**